README.md
# ghostscript-node

[![GitHub license](https://img.shields.io/github/license/sigalor/ghostscript-node)](https://github.com/sigalor/ghostscript-node/blob/master/LICENSE) [![npm](https://img.shields.io/npm/v/ghostscript-node)](https://www.npmjs.com/package/ghostscript-node) [![Unit tests workflow status](https://github.com/sigalor/ghostscript-node/actions/workflows/tests.yaml/badge.svg)](https://github.com/sigalor/ghostscript-node/actions/workflows/tests.yaml)

## Introduction

A fully promise-based Node.js library which can work with PDFs, based on Ghostscript and qpdf.

All PDF files that are handled by this library in the form of `Buffer` objects, i.e. to a user of this library, it looks like everything works only in memory. In the background though, file system access (via [tempy](https://www.npmjs.com/package/tempy)) is needed.

Additionally, this library requires the `gs` command (Ghostscript) as well as `qpdf` to be available. The required apt dependencies on Ubuntu 20.04 can be installed via the following command:

```
sudo apt install -y python python3 build-essential ghostscript libjpeg-dev libpng-dev libcurl4-openssl-dev mupdf-tools libfreetype6-dev qpdf
```

## Getting started

```
npm install ghostscript-node
```

## Usage

```javascript
const gs = require("ghostscript-node");

(async () => {
  const pdf1 = /* load first PDF as Buffer, e.g. from database */;
  const pdf2 = /* load second PDF as Buffer, e.g. from database */;

  // get Buffer object with bytes of combined PDF
  const combinedPDF = await gs.combinePDFs([pdf1, pdf2]);

  // get number of pages of pdf1
  const numPagesPDF1 = await gs.countPDFPages(pdf1);

  // get Buffer object with bytes of specified set of pages
  // page numbers begin with 1, last page is included
  const partOfPDF1 = await gs.extractPDFPages(pdf1, 3, 5);

  // rotate all pages of pdf1 by 90 degrees clockwise
  const rotatedPDF = await gs.rotatePDF(pdf1, "90");

  // returns an array of buffers containing PNG images of the desired pages
  const renderedPages = await gs.renderPDFPagesToPNG(pdf1);

  // checks if pdf1 is a valid PDF file
  const isPDF1Valid = await gs.isValidPDF(pdf1);
})();
```

## License

MIT


CHANGELOG.md
# 1.5.0 (2023-09-06)

- add `compressPDF` (gabolera) [#4](https://github.com/sigalor/ghostscript-node/pull/4)
- add `convertToPDFA` (MedAzizKhayati) [#6](https://github.com/sigalor/ghostscript-node/pull/6)

# 1.4.1 (2023-01-26)

- updated NPM dependencies

# 1.4.0 (2022-06-07)

- transformed entire library to use TypeScript
- fix `countPDFPages` to ignore Ghostscript warnings (e.g. due to a damaged XREF table)

# 1.3.2 (2022-02-16)

- escape input in countPDFPages for Windows, fix redundant await (Loupi) [#2](https://github.com/sigalor/ghostscript-node/pull/2)

# 1.3.1 (2022-01-24)

- fix handling of rotated PDFs
- improve internal structure of unit tests

# 1.3.0 (2022-01-06)

- added `isValidPDF` function

# 1.2.0 (2021-07-27)

- added `renderPDFPagesToPNG` function

# 1.1.2 (2021-01-28)

- use qpdf instead of PDFtk for rotating PDFs

# 1.1.1 (2021-01-28)

- fix TypeScript definition file

# 1.1.0 (2021-01-28)

- added `rotatePDF` function, based on PDFtk

# 1.0.0 (2021-01-25)

- initial release


test/index.test.js
const fs = require('fs-extra');
const gs = require('..');
const getFiles = require('./files');

let files;
beforeAll(async () => {
  files = await getFiles();
});

describe('combinePDFs', () => {
  test('works', async () => {
    const res = await gs.combinePDFs([files['pdf1.pdf'], files['pdf2.pdf']]);
    await expect(gs.countPDFPages(res)).resolves.toBe(4);
  });
});

describe('countPDFPages', () => {
  test('works', async () => {
    await expect(gs.countPDFPages(files['pdf1.pdf'])).resolves.toBe(1);
    await expect(gs.countPDFPages(files['pdf2.pdf'])).resolves.toBe(3);
  });

  test('fails for invalid PDF', async () => {
    await expect(gs.countPDFPages(Buffer.from([1, 2, 3]))).rejects.toThrow(
      /^Failed to determine number of pages in PDF: Command failed/,
    );
  });
});

describe('extractPDFPages', () => {
  test('works', async () => {
    const res = await gs.extractPDFPages(files['pdf2.pdf'], 2, 3);
    await expect(gs.countPDFPages(res)).resolves.toBe(2);
  });
});

describe('rotatePDF', () => {
  test('works', async () => {
    const res = await gs.rotatePDF(files['pdf2.pdf'], '90');
    await expect(gs.countPDFPages(res)).resolves.toBe(3);
  });
});

describe('renderPDFPagesToPNG', () => {
  test('works for single page PDF', async () => {
    const res = await gs.renderPDFPagesToPNG(files['pdf1.pdf']);
    expect(res.length).toBe(1);
  });

  test('works for three page PDF', async () => {
    const res = await gs.renderPDFPagesToPNG(files['pdf2.pdf']);
    expect(res.length).toBe(3);
  });

  test('works for second page of a PDF', async () => {
    const res = await gs.renderPDFPagesToPNG(files['pdf2.pdf'], 2, 2);
    expect(res.length).toBe(1);
  });

  test('works for the last two pages of a PDF', async () => {
    const res = await gs.renderPDFPagesToPNG(files['pdf2.pdf'], -2);
    expect(res.length).toBe(2);
  });

  test('fails when first page is 0', async () => {
    await expect(gs.renderPDFPagesToPNG(files['pdf1.pdf'], 0)).rejects.toThrow('First page number out of range: 0');
  });

  test('fails when last page is out of range', async () => {
    await expect(gs.renderPDFPagesToPNG(files['pdf1.pdf'], undefined, 3)).rejects.toThrow(
      'Last page number out of range: 3',
    );
  });

  test('fails when last page number is given when first page number is negative', async () => {
    await expect(gs.renderPDFPagesToPNG(files['pdf1.pdf'], -1, 2)).rejects.toThrow(
      'Last page must be undefined when first page is negative',
    );
  });

  test('fails for invalid PDF', async () => {
    await expect(gs.renderPDFPagesToPNG(Buffer.from([1, 2, 3]))).rejects.toThrow(
      /^Failed to determine number of pages in PDF: Command failed/,
    );
  });

  test('correctly renders rotated PDF', async () => {
    const rotatedPdf = await gs.rotatePDF(files['pdf1.pdf'], '90');
    const png = (await gs.renderPDFPagesToPNG(rotatedPdf))[0];
    expect(png.toString('base64')).toBe(files['renderPDFPagesToPNG-rotated-output.png'].toString('base64'));
  });
});

describe('isValidPDF', () => {
  test('returns true for valid PDF', async () => {
    expect(await gs.isValidPDF(files['pdf1.pdf'])).toBe(true);
  });

  test('returns false for invalid PDF', async () => {
    expect(await gs.isValidPDF(Buffer.from([1, 2, 3]))).toBe(false);
  });
});


test/index.test.ts
import * as gs from '../src/index';
import { FilesMap, getFiles } from './files';

let files: FilesMap;
beforeAll(async () => {
  files = await getFiles();
});

describe('combinePDFs', () => {
  test('works', async () => {
    const res = await gs.combinePDFs([files['pdf1.pdf'], files['pdf2.pdf']]);
    await expect(gs.countPDFPages(res)).resolves.toBe(4);
  });
});

describe('countPDFPages', () => {
  test('works', async () => {
    await expect(gs.countPDFPages(files['pdf1.pdf'])).resolves.toBe(1);
    await expect(gs.countPDFPages(files['pdf2.pdf'])).resolves.toBe(3);
  });

  test('fails for invalid PDF', async () => {
    await expect(gs.countPDFPages(Buffer.from([1, 2, 3]))).rejects.toThrow(
      /^Failed to determine number of pages in PDF: Command failed/,
    );
  });
});

describe('extractPDFPages', () => {
  test('works', async () => {
    const res = await gs.extractPDFPages(files['pdf2.pdf'], 2, 3);
    await expect(gs.countPDFPages(res)).resolves.toBe(2);
  });
});

describe('rotatePDF', () => {
  test('works', async () => {
    const res = await gs.rotatePDF(files['pdf2.pdf'], '90');
    await expect(gs.countPDFPages(res)).resolves.toBe(3);
  });
});

describe('convertToPDFA', () => {
  test('works', async () => {
    const res = await gs.convertToPDFA(files['pdf2.pdf']);
    await expect(gs.countPDFPages(res)).resolves.toBe(3);
  });
});

describe('renderPDFPagesToPNG', () => {
  test('works for single page PDF', async () => {
    const res = await gs.renderPDFPagesToPNG(files['pdf1.pdf']);
    expect(res.length).toBe(1);
  });

  test('works for three page PDF', async () => {
    const res = await gs.renderPDFPagesToPNG(files['pdf2.pdf']);
    expect(res.length).toBe(3);
  });

  test('works for second page of a PDF', async () => {
    const res = await gs.renderPDFPagesToPNG(files['pdf2.pdf'], 2, 2);
    expect(res.length).toBe(1);
  });

  test('works for the last two pages of a PDF', async () => {
    const res = await gs.renderPDFPagesToPNG(files['pdf2.pdf'], -2);
    expect(res.length).toBe(2);
  });

  test('fails when first page is 0', async () => {
    await expect(gs.renderPDFPagesToPNG(files['pdf1.pdf'], 0)).rejects.toThrow('First page number out of range: 0');
  });

  test('fails when last page is out of range', async () => {
    await expect(gs.renderPDFPagesToPNG(files['pdf1.pdf'], undefined, 3)).rejects.toThrow(
      'Last page number out of range: 3',
    );
  });

  test('fails when last page number is given when first page number is negative', async () => {
    await expect(gs.renderPDFPagesToPNG(files['pdf1.pdf'], -1, 2)).rejects.toThrow(
      'Last page must be undefined when first page is negative',
    );
  });

  test('fails for invalid PDF', async () => {
    await expect(gs.renderPDFPagesToPNG(Buffer.from([1, 2, 3]))).rejects.toThrow(
      /^Failed to determine number of pages in PDF: Command failed/,
    );
  });

  test('correctly renders rotated PDF', async () => {
    const rotatedPdf = await gs.rotatePDF(files['pdf1.pdf'], '90');
    const png = (await gs.renderPDFPagesToPNG(rotatedPdf))[0];
    expect([
      files['renderPDFPagesToPNG-rotated-output.png'].toString('base64'),
      files['renderPDFPagesToPNG-rotated-output2.png'].toString('base64'),
    ]).toContain(png.toString('base64'));
  });
});

describe('isValidPDF', () => {
  test('returns true for valid PDF', async () => {
    expect(await gs.isValidPDF(files['pdf1.pdf'])).toBe(true);
  });

  test('returns true for another valid PDF', async () => {
    expect(await gs.isValidPDF(files['pdf3.pdf'])).toBe(true);
  });

  test('returns true for PDF that has been extracted and combined again', async () => {
    const page = await gs.extractPDFPages(files['pdf3.pdf'], 1, 1);
    const pdf = await gs.combinePDFs([page]);
    expect(await gs.isValidPDF(pdf)).toBe(true);
  });

  test('returns false for invalid PDF', async () => {
    expect(await gs.isValidPDF(Buffer.from([1, 2, 3]))).toBe(false);
  });
});


describe('compressPDF', () => {
  test('returns PDF reduce size send buffer file', async () => {
    const optimizedPDF = await gs.compressPDF(files['pdf3.pdf'])
    expect(optimizedPDF.length).toBeLessThanOrEqual(files['pdf3.pdf'].length);
  });
  test('returns PDF reduce size send string encoded file', async () => {
    const optimizedPDF = await gs.compressPDF(files['pdf3.pdf'].toString('base64'))
    expect(optimizedPDF.length).toBeLessThanOrEqual(files['pdf3.pdf'].length);
  });
})

test/files/index.js
const fs = require('fs-extra');
const path = require('path');

module.exports = async () => {
  const fileNames = (await fs.readdir(__dirname)).filter(f => f !== 'index.js');
  const files = {};

  for (let fileName of fileNames) {
    files[fileName] = await fs.readFile(path.join(__dirname, fileName));
  }

  return files;
};


index.js
const childProcess = require('child_process');
const fs = require('fs-extra');
const tempy = require('tempy');
const util = require('util');
const exec = util.promisify(childProcess.exec);

async function useTempFiles(filenameSets, fn) {
  const filenames = {};

  // create all desired temporary files (either empty files or write data buffers to them)
  for (const [k, config] of Object.entries(filenameSets)) {
    const { numFiles, writeBuffers, ...tempyConfig } = config;

    if (numFiles !== undefined) {
      filenames[k] = Array(numFiles)
        .fill(0)
        .map(() => tempy.file(tempyConfig));
    } else if (writeBuffers !== undefined) {
      filenames[k] = await Promise.all(writeBuffers.map(b => tempy.write(b, tempyConfig)));
    }
  }

  // execute the worker function which uses these files
  const ret = await fn(filenames);

  // remove all the temporary files again
  await Promise.all([].concat(...Object.values(filenames)).map(f => fs.unlink(f)));

  // return the result of the worker function
  return ret;
}

async function useTempFilesPDF(filenameSets, fn) {
  Object.values(filenameSets).forEach(v => (v.extension = '.pdf'));
  return useTempFiles(filenameSets, fn);
}

// writes inputBuffer to one temporary file, creates an empty output file and eventually returns output
async function useTempFilesPDFInOut(inputBuffer, fn) {
  return useTempFilesPDF(
    { input: { writeBuffers: [inputBuffer] }, output: { numFiles: 1 } },
    async ({ input, output }) => {
      await fn(input[0], output[0]);
      return fs.readFile(output[0]);
    },
  );
}

async function useTempFilesPDFIn(inputBuffer, fn) {
  return useTempFilesPDF({ input: { writeBuffers: [inputBuffer] } }, async ({ input }) => fn(input[0]));
}

async function combinePDFs(pdfBuffers) {
  if (pdfBuffers.length === 0) return Buffer.alloc(0);
  if (pdfBuffers.length === 1) return pdfBuffers[0];

  try {
    return await useTempFilesPDF(
      { inputs: { writeBuffers: pdfBuffers }, output: { numFiles: 1 } },
      async ({ inputs, output }) => {
        await exec(
          `gs -q -dNOPAUSE -sDEVICE=pdfwrite -sOUTPUTFILE=${output[0]} -dBATCH -dAutoRotatePages=/None ${inputs.join(
            ' ',
          )} -c "[ /Creator () /Producer () /DOCINFO pdfmark"`,
        );
        return fs.readFile(output[0]);
      },
    );
  } catch (e) {
    throw new Error('Failed to combine PDFs: ' + e.message);
  }
}

async function countPDFPages(pdfBuffer) {
  try {
    return await useTempFilesPDFIn(pdfBuffer, async input => {
      const escapedInput = input.replace(/\\/g, '\\\\')
      const { stdout } = await exec(
        `gs -q -dNOPAUSE -dBATCH -dNOSAFER -dNODISPLAY -c "(${escapedInput}) (r) file runpdfbegin pdfpagecount = quit"`,
      );
      return parseInt(stdout);
    });
  } catch (e) {
    throw new Error('Failed to determine number of pages in PDF: ' + e.message);
  }
}

async function extractPDFPages(pdfBuffer, firstPage, lastPage) {
  try {
    return await useTempFilesPDFInOut(pdfBuffer, async (input, output) => {
      await exec(
        `gs -q -dNOPAUSE -sDEVICE=pdfwrite -dBATCH -dNOSAFER -dFirstPage=${firstPage} -dLastPage=${lastPage} -dAutoRotatePages=/None -sOutputFile=${output} ${input}`,
      );
    });
  } catch (e) {
    throw new Error('Failed to extract PDF pages: ' + e.message);
  }
}

async function rotatePDF(pdfBuffer, direction) {
  if (!['90', '180', '270'].includes(direction)) throw new Error('Invalid rotation direction: ' + direction);

  try {
    return await useTempFilesPDFInOut(pdfBuffer, async (input, output) => {
      await exec(`qpdf ${input} ${output} --rotate=${direction}`);
    });
  } catch (e) {
    throw new Error('Failed to rotate PDF: ' + e.message);
  }
}

/**
 * If `firstPage` is not given, 1 is used.
 * If `lastPage` is not given, the document's last page is used.
 * If `firstPage` is negative (e.g. -n), this refers to the last n pages and `lastPage` must be undefined.
 * All page numbers start at 1.
 */
async function renderPDFPagesToPNG(pdfBuffer, firstPage, lastPage, resolution = 300) {
  const numPages = await countPDFPages(pdfBuffer);

  if (firstPage === undefined) firstPage = 1;
  else if (firstPage === 0 || (firstPage < 0 && firstPage < -numPages))
    throw new Error('First page number out of range: ' + firstPage);

  if (firstPage < 0) {
    if (lastPage !== undefined) throw new Error('Last page must be undefined when first page is negative');
    firstPage = numPages + firstPage + 1;
    lastPage = numPages;
  }

  if (lastPage === undefined) lastPage = numPages;
  else if (lastPage > numPages) throw new Error('Last page number out of range: ' + lastPage);

  if (firstPage > lastPage) throw new Error('Invalid page range: ' + firstPage + '-' + lastPage);

  try {
    return await useTempFilesPDFIn(pdfBuffer, async input => {
      const outDir = tempy.directory();
      await exec(
        `gs -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 -sDEVICE=png16m -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -r${resolution} -sOutputFile=${outDir}/%d.png -dFirstPage=${firstPage} -dLastPage=${lastPage} ${input}`,
      );

      const outFiles = [];
      for (let i = 1; i <= lastPage - firstPage + 1; i++) {
        outFiles.push(await fs.readFile(outDir + '/' + i + '.png'));
      }

      await fs.rmdir(outDir, { recursive: true });
      return outFiles;
    });
  } catch (e) {
    throw new Error('Failed to render PDF pages to PNG: ' + e.message);
  }
}

async function isValidPDF(pdfBuffer) {
  try {
    await countPDFPages(pdfBuffer);
    return true;
  } catch (e) {
    return false;
  }
}

module.exports = {
  combinePDFs,
  countPDFPages,
  extractPDFPages,
  rotatePDF,
  renderPDFPagesToPNG,
  isValidPDF,
};


src/index.ts
import childProcess from 'child_process';
import fs from 'fs-extra';
import tempy, { FileOptions } from 'tempy';
import util from 'util';

const exec = util.promisify(childProcess.exec);

interface TempFilenameSetDefs {
  [setName: string]: {
    writeBuffers?: Buffer[];
    numFiles?: number;
    tempyConfig?: FileOptions;
  };
}

type TempFilenameSets = {
  [setName: string]: string[];
};

type TempFileFnSingle<T> = (input: string, output?: string) => Promise<T>;
type TempFileFnMany<T> = (filenames: TempFilenameSets) => Promise<T>;

async function useTempFiles<T>(filenameSets: TempFilenameSetDefs, fn: TempFileFnMany<T>) {
  const filenames: TempFilenameSets = {};

  // create all desired temporary files (either empty files or write data buffers to them)
  for (const [k, config] of Object.entries(filenameSets)) {
    const { numFiles, writeBuffers, tempyConfig } = config;

    if (numFiles !== undefined) {
      filenames[k] = Array(numFiles)
        .fill(0)
        .map(() => tempy.file(tempyConfig));
    } else if (writeBuffers !== undefined) {
      filenames[k] = await Promise.all(writeBuffers.map(b => tempy.write(b, tempyConfig)));
    }
  }

  // execute the worker function which uses these files
  const ret = await fn(filenames);

  // remove all the temporary files again
  await Promise.all([].concat(...(<any>Object.values(filenames))).map(f => fs.unlink(f)));

  // return the result of the worker function
  return ret;
}

async function useTempFilesPDF<T>(filenameSets: TempFilenameSetDefs, fn: TempFileFnMany<T>) {
  Object.values(filenameSets).forEach(v => (v.tempyConfig = { extension: '.pdf' }));
  return useTempFiles(filenameSets, fn);
}

// writes inputBuffer to one temporary file, creates an empty output file, calls the worker function and returns the output file contents as Buffer
async function useTempFilesPDFInOut(inputBuffer: Buffer, fn: TempFileFnSingle<void>): Promise<Buffer> {
  return useTempFilesPDF(
    { input: { writeBuffers: [inputBuffer] }, output: { numFiles: 1 } },
    async ({ input, output }) => {
      await fn(input[0], output[0]);
      return fs.readFile(output[0]);
    },
  );
}

async function useTempFilesPDFIn<T>(inputBuffer: Buffer, fn: TempFileFnSingle<T>): Promise<T> {
  return useTempFilesPDF({ input: { writeBuffers: [inputBuffer] } }, async ({ input }) => fn(input[0]));
}

export async function combinePDFs(pdfBuffers: Buffer[]): Promise<Buffer> {
  if (pdfBuffers.length === 0) return Buffer.alloc(0);
  if (pdfBuffers.length === 1) return pdfBuffers[0];

  try {
    return await useTempFilesPDF(
      { inputs: { writeBuffers: pdfBuffers }, output: { numFiles: 1 } },
      async ({ inputs, output }) => {
        await exec(
          `gs -q -dNOPAUSE -sDEVICE=pdfwrite -sOUTPUTFILE=${output[0]} -dBATCH -dAutoRotatePages=/None ${inputs.join(
            ' ',
          )} -c "[ /Creator () /Producer () /DOCINFO pdfmark"`,
        );
        return fs.readFile(output[0]);
      },
    );
  } catch (e: any) {
    throw new Error('Failed to combine PDFs: ' + e.message);
  }
}

export async function countPDFPages(pdfBuffer: Buffer): Promise<number> {
  try {
    return await useTempFilesPDFIn<number>(pdfBuffer, async input => {
      const escapedInput = input.replace(/\\/g, '\\\\');
      let { stdout } = await exec(
        `gs -q -dNOPAUSE -dBATCH -dNOSAFER -dNODISPLAY -c "(${escapedInput}) (r) file runpdfbegin pdfpagecount = quit"`,
      );

      /**
       * Necessary for when Ghostscript detects a damaged but repairable PDF file, because then it outputs
       * the following before the number of pages (three spaces before each line starting with "****"):
       *
       *    **** Error:  An error occurred while reading an XREF table.
       *    **** The file has been damaged.  This may have been caused
       *    **** by a problem while converting or transfering the file.
       *    **** Ghostscript will attempt to recover the data.
       *    **** However, the output may be incorrect.
       */
      stdout = stdout
        .split('\n')
        .filter(l => !l.startsWith('   **** ') && l.length > 0)
        .join('\n')
        .trim();

      const ret = parseInt(stdout);
      if (isNaN(ret)) throw new Error('parsing failed: ' + stdout);
      return ret;
    });
  } catch (e: any) {
    throw new Error('Failed to determine number of pages in PDF: ' + e.message);
  }
}

export async function extractPDFPages(pdfBuffer: Buffer, firstPage: number, lastPage: number): Promise<Buffer> {
  try {
    return await useTempFilesPDFInOut(pdfBuffer, async (input, output) => {
      await exec(
        `gs -q -dNOPAUSE -sDEVICE=pdfwrite -dBATCH -dNOSAFER -dFirstPage=${firstPage} -dLastPage=${lastPage} -dAutoRotatePages=/None -sOutputFile=${output} ${input}`,
      );
    });
  } catch (e: any) {
    throw new Error('Failed to extract PDF pages: ' + e.message);
  }
}

export async function rotatePDF(pdfBuffer: Buffer, direction: '90' | '180' | '270'): Promise<Buffer> {
  if (!['90', '180', '270'].includes(direction)) throw new Error('Invalid rotation direction: ' + direction);

  try {
    return await useTempFilesPDFInOut(pdfBuffer, async (input, output) => {
      await exec(`qpdf ${input} ${output} --rotate=${direction}`);
    });
  } catch (e: any) {
    throw new Error('Failed to rotate PDF: ' + e.message);
  }
}

/**
 * Converts a PDF to PDF/A.
 * 
 * @param pdfBuffer - Buffer of the PDF to convert
 * @param options - Options for the conversion
 * @param options.version - PDF/A version to convert to. Defaults to 1.
 * @returns 
 */
export async function convertToPDFA(
  pdfBuffer: Buffer,
  options = {
    version: 1,
  },
): Promise<Buffer> {
  try {
    return await useTempFilesPDFInOut(pdfBuffer, async (input, output) => {
      await exec(
        `gs -dPDFA -dBATCH -dNOPAUSE -sColorConversionStrategy=UseDeviceIndependentColor -sDEVICE=pdfwrite -dPDFACompatibilityPolicy=${options.version} -sOutputFile=${output} ${input}`,
      );
    });
  } catch (e: any) {
    throw new Error('Failed to convert PDF to PDF/A: ' + e.message);
  }
}

/**
 * If `firstPage` is not given, 1 is used.
 * If `lastPage` is not given, the document's last page is used.
 * If `firstPage` is negative (e.g. -n), this refers to the last n pages and `lastPage` must be undefined.
 * All page numbers start at 1.
 */
export async function renderPDFPagesToPNG(
  pdfBuffer: Buffer,
  firstPage?: number,
  lastPage?: number,
  resolution = 300,
): Promise<Buffer[]> {
  const numPages = await countPDFPages(pdfBuffer);

  if (firstPage === undefined) firstPage = 1;
  else if (firstPage === 0 || (firstPage < 0 && firstPage < -numPages))
    throw new Error('First page number out of range: ' + firstPage);

  if (firstPage < 0) {
    if (lastPage !== undefined) throw new Error('Last page must be undefined when first page is negative');
    firstPage = numPages + firstPage + 1;
    lastPage = numPages;
  }

  if (lastPage === undefined) lastPage = numPages;
  else if (lastPage > numPages) throw new Error('Last page number out of range: ' + lastPage);

  if (firstPage! > lastPage!) throw new Error('Invalid page range: ' + firstPage + '-' + lastPage);

  try {
    return await useTempFilesPDFIn(pdfBuffer, async input => {
      const outDir = tempy.directory();
      await exec(
        `gs -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 -sDEVICE=png16m -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -r${resolution} -sOutputFile=${outDir}/%d.png -dFirstPage=${firstPage} -dLastPage=${lastPage} ${input}`,
      );

      const outFiles = [];
      for (let i = 1; i <= lastPage! - firstPage! + 1; i++) {
        outFiles.push(await fs.readFile(outDir + '/' + i + '.png'));
      }

      await fs.rmdir(outDir, { recursive: true });
      return outFiles;
    });
  } catch (e: any) {
    throw new Error('Failed to render PDF pages to PNG: ' + e.message);
  }
}

export async function isValidPDF(pdfBuffer: Buffer): Promise<boolean> {
  try {
    await countPDFPages(pdfBuffer);
    return true;
  } catch (e) {
    return false;
  }
}

/**
 * This function try, reduce size of your PDF not destroying quality
 * @param pdfBuffer Buffer
 * @returns Buffer
 */
export async function compressPDF(pdfBuffer: Buffer | string, encoding?: BufferEncoding): Promise<Buffer> {
  try {
    if(typeof pdfBuffer === 'string'){
      pdfBuffer = Buffer.from(pdfBuffer, encoding ?? 'base64')
    }
    const compressedPdf = await useTempFilesPDFInOut(pdfBuffer, async (input, output) => {
      await exec(
        `gs -q -dNOPAUSE -dBATCH -dSAFER -sDEVICE=pdfwrite -dCompatibilityLevel=1.4 -dPDFSETTINGS=/screen -dEmbedAllFonts=true -dSubsetFonts=true -dColorImageDownsampleType=/Bicubic -dColorImageResolution=144 -dGrayImageDownsampleType=/Bicubic -dGrayImageResolution=144 -dMonoImageDownsampleType=/Bicubic -dMonoImageResolution=144 -sOutputFile=${output} ${input}`,
      );
    });
    if (pdfBuffer.length < compressedPdf.length) {
      return pdfBuffer;
    }
    return compressedPdf;
  } catch (e: any) {
    throw new Error('Failed optimize PDF: ' + e.message);
  }
}
